### ISSUE TITLE
Clarification on the inception of a new document's ID

### ISSUE THREAD

**Issue by llorllale** [decentralized-identity/secure-data-store#112]
We have implemented an [encrypted data vault](https://github.com/trustbloc/edv) and we are now moving to integrate an EDV client for it as a "storage plugin" to [aries-framework-go](https://github.com/hyperledger/aries-framework-go).

One of the sticking points is the ambiguity in the spec w.r.t. a new document's identifier.

To create a new document one must `POST` a request with an EncryptedDocument structure. This structure has an `id` property that the client assigns a value to. Inception of this ID is on the client side.

[This example response](https://identity.foundation/secure-data-store/#example-10-successful-data-vault-document-creation-response) is returning a different ID than the one sent in the example request right above it.

**Request**

* Clarification on where the spec or the group's thinking is headed in, whether the identifier is controlled by the client or controlled by the server, and
* Add normative text to the spec that clarifies this point

**Comment by llorllale** [comment]
related: #98

**Comment by dlongley** [comment]
I think the example response is in error.

My thinking is that the ID must be controlled by the client. This is in order to have stable IDs for documents no matter which storage server is used. If you wanted to migrate to another storage server and it rewrote all of your IDs, that would be problematic. Therefore, IDs should be generated by the client and follow a specific profile so that all compliant servers will accept them. The profile suggested is a 128-bit multibase-encoded (base58) random value. This allows servers to decode IDs and store them in binary (if desired) and it is a sufficiently large enough number such that there should not be conflicts.

+1 for normative text clarifying this point and the reasons behind it.

**Comment by llorllale** [comment]
@dlongley you beat me to it.

> My thinking is that the ID must be controlled by the client.

I agree.

It is very convenient and cheap to be able to use a sufficiently unique identifier that two or more parties can refer to (such as a DID) as the key for this document in the EDV. Having the server generate the ID for the document means the client must now either add an index for it in the EncryptedDocument, or shoulder the burden of having to maintain a separate mapping table with related metadata.

It seems to me we would see little benefit in having the EDV server generating the IDs, but we would have to deal with non-trivial complexity and performance considerations.

**Comment by llorllale** [comment]
@dlongley 
(emphasis mine)

> The profile **suggested** is a 128-bit multibase-encoded (base58) random value.

The text seems normative and is a MUST:

> An identifier for the structured document. The value is required and MUST be a Base58-encoded 128-bit random value.

**Comment by dlongley** [comment]
@llorllale,

Yes, I meant that is the profile suggested *by me* (and others who have worked on the current implementations). I agree that whatever the profile is (preferably it will be what I wrote :smiley:) is normative in the spec.

**Comment by dlongley** [comment]
To be more specific, I think it should be multibase encoded to support future proofing:

> An identifier for the structured document. The value is required and MUST be a multibase, Base58-encoded 128-bit random value.

**Comment by llorllale** [comment]
@dlongley @dmitrizagidulin 

We have a use case for range queries with a prefix. The prefix is not known ahead of time. Any randomization of the document IDs would not work for us.

What is the reason for the insistence of randomization?

**Comment by Baha-sk** [comment]
@dlongley @llorllale since a document is a JWE envelope, we know it is controlled by the recipient key. And since it's an asymmetric key, it would be easy to compute it's JWK fingerprint (which is a non-random deterministic value) appended with a numeric id as suffix (to represent multiple documents using the same key).

This way, a client with the recipient key can easily build the ID and a potential storage iterator only needs to iterate through the incremental numeric suffix.

**Comment by llorllale** [comment]
@Baha-sk we do not wish to use the jwe recipient key(s) as identifiers for the documents.

**Comment by Baha-sk** [comment]
Unless we have a deterministic way to iterate through documents, a storage cannot range query documents. 

At minimum, we should have such kind of index, probably not the main ID of a document, but we definitely need an additional index for range query.

**Comment by llorllale** [comment]
@dlongley @dmitrizagidulin thoughts?

**Comment by dlongley** [comment]
@llorllale,

> We have a use case for range queries with a prefix. The prefix is not known ahead of time. Any randomization of the document IDs would not work for us.
> 
> What is the reason for the insistence of randomization?

To minimize the knowledge of the storage server as much as possible. If IDs contain semantic information, that's additional information the storage server knows about your data.

We should discuss the range query use case in the group and see what we can come up with as a solution.
